src/components/kpi-group/index.tsx

```diff
@@ -34,7 +34,7 @@ export class KpiGroup extends React.Component<KpiGroupProps> {
 
     private renderKpi = (kpi: KpiData, showBorder: boolean = true) => {
         return (
-            <div key={kpi.name} className={'kpi-group_stat-wrapper'}>
+            <div key={kpi.name + kpi.name + kpi.subvalue + kpi.inlineValue} className={'kpi-group_stat-wrapper'}>

```

Comment:
Why `kpi.name + kpi.name`? Suggest de-duplicating it
```suggestion
            <div key={kpi.name + kpi.subvalue + kpi.inlineValue} className={'kpi-group_stat-wrapper'}>
```

Reply:
Can you elaborate a bit more on why this is necessary or why you think this is the best fix? The idea behind a key in react is so that if you have a list of items, and the value of one of the items changes, only that element needs to be re-rendered. Or if an item is removed from the list, react can remove the item whose key no longer appears in the list without comparing the actual content. So for example, if there was a list of scores for players A, B, and C, if the key was the name of the player, then when the scores changed from A: 100, B: 200, C: 300 to A: 100, B: 400, C: 300, the elements for A, B, and C would not need to be destroyed and recreated, the only change in the DOM would be changing the text for B to 400.

So in this case, I was simply using the "name" as the key because in most of my kpis, the name was the difficulty or the player or something, and each instance of the kpi in a card was for a different difficulty or a different player.

I can see now that this is flawed because in your use case, the kpis might have the same name.

But given my explanation above, do you think that combining the name, subvalue, and inline value is the best way to make a key?

In the example above, changing the score would change the key, so react wouldn't know to reuse the element it previously used for B. It _might_ reuse it, but that would be more by chance/convenience.

So is there something else that can be done within the kpi group to create a key that is tied to the "identity" of a kpi without being tied to the values which might change?

**Note:** In the grand scheme of things, this key is fine, I'm just trying to get both of us to think about better ways if we can :-)

Reply:
discussed offline: keys should come from outside